README.txt
ljdump - livejournal archiver

This program reads the journal entries from a livejournal (or compatible)
blog site and archives them in a subdirectory named after the journal name.

The simplest way to run this is to execute the ljdump.py script with Python.
Depending on your OS, you may be able to double-click the ljdump.py script
directly, or you may need to open a Terminal/Command Prompt window to run it.
Either way, it will prompt you for your Livejournal username and password,
then download all your journal entries, comments, and userpics.

You may optionally download entries from a different journal (a community)
where you are a member. If you are a community maintainer, you can also
download comments from the community.

If you want to save your username and password so you don't have to type
it every time you run ljdump, you can save it in the configuration file.

The configuration is read from "ljdump.config". A sample configuration is
provided in "ljdump.config.sample", which should be copied and then edited.
The configuration settings are:

  server - The XMLRPC server URL. This should only need to be changed
           if you are dumping a journal that is livejournal-compatible
           but is not livejournal itself.

  username - The livejournal user name. A subdirectory will be created
             with this same name to store the journal entries.

  password - The account password. This password is never sent in the
             clear; the livejournal "challenge" password mechanism is used.

  journal - Optional: The journal to download entries from. If this is
            not specified, the "username" journal is downloaded. If this
            is specified, then only the named journals will be downloaded
            (this element may be specified more than once to download
            multiple journals).

This program may be run as often as needed to bring the backup copy up
to date. Both new and updated items are downloaded.

The community http://ljdump.livejournal.com has been set up for questions
or comments.

-----

convertdump - convert ljdump format to wordpress importer format
Contributed by Sean M. Graham (www.sean-graham.com) and others

This will convert a pydump archive into something compatible with the
WordPress LiveJournal importer.  This is the same format used by the Windows
ljArchive exporter.

Arguments:
    -u  --user      username of archive to process [required]
    -l  --limit     limit the number of entries in each xml file (default 250)
    -i  --insecure  include private and protected entries in the output
    -h  --help      show this help page

Example:
    ./convertdump.py --user stevemartin --limit 200 --insecure


ChangeLog
ChangeLog - ljdump

Version 1.5.1 - 2010-12-29

- Suppress warning about deprecated md5 module
- Don't attempt to fetch default userpic if the user hasn't set one

Version 1.5 - 2009-02-04

- Include convertdump.py for Wordpress conversions.

Version 1.4 - 2009-01-13

- Now supports downloading entries and comments from communities

Version 1.3.3 - 2009-01-10

- Feature: ljdump now prompts for login info if ljdump.config does not exist
- Bugfix: Automatically handle UnicodeDecodeError for old entries
- Bugfix: Handle more non-filename characters of userpic keywords

Version 1.3.2 - 2009-01-09

- Bugfix: Handle case where userpic keywords contain a / or \.

Version 1.3.1 - 2009-01-08

- Feature: Download and save userpic images
- Feature: Resume download where it left off if interrupted
- Bugfix: Handle userpic keywords with non-ASCII characters
- Bugfix: Handle unexpected empty item

Version 1.2 - 2006-09-08

- Fetch userpic data into userpics.xml

Version 1.1 - 2006-01-09

- Add comment downloading.

Version 1.0.3 - 2005-12-25

- Get updated journal entries in addition to new entries

Version 1.0.2 - 2005-11-25

- Set +x on ljdump.py

Version 1.0.1 - 2005-11-25

- Corrected total items count
- Added shebang line to top of script

Version 1.0 - 2005-11-25

- Initial version


ljdump.py
#!/usr/bin/python
#
# ljdump.py - livejournal archiver
# Greg Hewgill <greg@hewgill.com> http://hewgill.com
# Version 1.5.1
#
# LICENSE
#
# This software is provided 'as-is', without any express or implied
# warranty.  In no event will the author be held liable for any damages
# arising from the use of this software.
#
# Permission is granted to anyone to use this software for any purpose,
# including commercial applications, and to alter it and redistribute it
# freely, subject to the following restrictions:
#
# 1. The origin of this software must not be misrepresented; you must not
#    claim that you wrote the original software. If you use this software
#    in a product, an acknowledgment in the product documentation would be
#    appreciated but is not required.
# 2. Altered source versions must be plainly marked as such, and must not be
#    misrepresented as being the original software.
# 3. This notice may not be removed or altered from any source distribution.
#
# Copyright (c) 2005-2010 Greg Hewgill and contributors

import codecs, os, pickle, pprint, re, shutil, sys, urllib2, xml.dom.minidom, xmlrpclib
from xml.sax import saxutils

MimeExtensions = {
    "image/gif": ".gif",
    "image/jpeg": ".jpg",
    "image/png": ".png",
}

try:
    from hashlib import md5
except ImportError:
    import md5 as _md5
    md5 = _md5.new

def calcchallenge(challenge, password):
    return md5(challenge+md5(password).hexdigest()).hexdigest()

def flatresponse(response):
    r = {}
    while True:
        name = response.readline()
        if len(name) == 0:
            break
        if name[-1] == '\n':
            name = name[:len(name)-1]
        value = response.readline()
        if value[-1] == '\n':
            value = value[:len(value)-1]
        r[name] = value
    return r

def getljsession(server, username, password):
    r = urllib2.urlopen(server+"/interface/flat", "mode=getchallenge")
    response = flatresponse(r)
    r.close()
    r = urllib2.urlopen(server+"/interface/flat", "mode=sessiongenerate&user=%s&auth_method=challenge&auth_challenge=%s&auth_response=%s" % (username, response['challenge'], calcchallenge(response['challenge'], password)))
    response = flatresponse(r)
    r.close()
    return response['ljsession']

def dochallenge(server, params, password):
    challenge = server.LJ.XMLRPC.getchallenge()
    params.update({
        'auth_method': "challenge",
        'auth_challenge': challenge['challenge'],
        'auth_response': calcchallenge(challenge['challenge'], password)
    })
    return params

def dumpelement(f, name, e):
    f.write("<%s>\n" % name)
    for k in e.keys():
        if isinstance(e[k], {}.__class__):
            dumpelement(f, k, e[k])
        else:
            try:
                s = unicode(str(e[k]), "UTF-8")
            except UnicodeDecodeError:
                # fall back to Latin-1 for old entries that aren't UTF-8
                s = unicode(str(e[k]), "cp1252")
            f.write("<%s>%s</%s>\n" % (k, saxutils.escape(s), k))
    f.write("</%s>\n" % name)

def writedump(fn, event):
    f = codecs.open(fn, "w", "UTF-8")
    f.write("""<?xml version="1.0"?>\n""")
    dumpelement(f, "event", event)
    f.close()

def writelast(journal, lastsync, lastmaxid):
    f = open("%s/.last" % journal, "w")
    f.write("%s\n" % lastsync)
    f.write("%s\n" % lastmaxid)
    f.close()

def createxml(doc, name, map):
    e = doc.createElement(name)
    for k in map.keys():
        me = doc.createElement(k)
        me.appendChild(doc.createTextNode(map[k]))
        e.appendChild(me)
    return e

def gettext(e):
    if len(e) == 0:
        return ""
    return e[0].firstChild.nodeValue

def ljdump(Server, Username, Password, Journal):
    m = re.search("(.*)/interface/xmlrpc", Server)
    if m:
        Server = m.group(1)
    if Username != Journal:
        authas = "&authas=%s" % Journal
    else:
        authas = ""

    print "Fetching journal entries for: %s" % Journal
    try:
        os.mkdir(Journal)
        print "Created subdirectory: %s" % Journal
    except:
        pass

    ljsession = getljsession(Server, Username, Password)

    server = xmlrpclib.ServerProxy(Server+"/interface/xmlrpc")

    newentries = 0
    newcomments = 0
    errors = 0

    lastsync = ""
    lastmaxid = 0
    try:
        f = open("%s/.last" % Journal, "r")
        lastsync = f.readline()
        if lastsync[-1] == '\n':
            lastsync = lastsync[:len(lastsync)-1]
        lastmaxid = f.readline()
        if len(lastmaxid) > 0 and lastmaxid[-1] == '\n':
            lastmaxid = lastmaxid[:len(lastmaxid)-1]
        if lastmaxid == "":
            lastmaxid = 0
        else:
            lastmaxid = int(lastmaxid)
        f.close()
    except:
        pass
    origlastsync = lastsync

    r = server.LJ.XMLRPC.login(dochallenge(server, {
        'username': Username,
        'ver': 1,
        'getpickws': 1,
        'getpickwurls': 1,
    }, Password))
    userpics = dict(zip(map(str, r['pickws']), r['pickwurls']))
    if r['defaultpicurl']:
        userpics['*'] = r['defaultpicurl']

    while True:
        r = server.LJ.XMLRPC.syncitems(dochallenge(server, {
            'username': Username,
            'ver': 1,
            'lastsync': lastsync,
            'usejournal': Journal,
        }, Password))
        #pprint.pprint(r)
        if len(r['syncitems']) == 0:
            break
        for item in r['syncitems']:
            if item['item'][0] == 'L':
                print "Fetching journal entry %s (%s)" % (item['item'], item['action'])
                try:
                    e = server.LJ.XMLRPC.getevents(dochallenge(server, {
                        'username': Username,
                        'ver': 1,
                        'selecttype': "one",
                        'itemid': item['item'][2:],
                        'usejournal': Journal,
                    }, Password))
                    if e['events']:
                        writedump("%s/%s" % (Journal, item['item']), e['events'][0])
                        newentries += 1
                    else:
                        print "Unexpected empty item: %s" % item['item']
                        errors += 1
                except xmlrpclib.Fault, x:
                    print "Error getting item: %s" % item['item']
                    pprint.pprint(x)
                    errors += 1
            lastsync = item['time']
            writelast(Journal, lastsync, lastmaxid)

    # The following code doesn't work because the server rejects our repeated calls.
    # http://www.livejournal.com/doc/server/ljp.csp.xml-rpc.getevents.html
    # contains the statement "You should use the syncitems selecttype in
    # conjuntions [sic] with the syncitems protocol mode", but provides
    # no other explanation about how these two function calls should
    # interact. Therefore we just do the above slow one-at-a-time method.

    #while True:
    #    r = server.LJ.XMLRPC.getevents(dochallenge(server, {
    #        'username': Username,
    #        'ver': 1,
    #        'selecttype': "syncitems",
    #        'lastsync': lastsync,
    #    }, Password))
    #    pprint.pprint(r)
    #    if len(r['events']) == 0:
    #        break
    #    for item in r['events']:
    #        writedump("%s/L-%d" % (Journal, item['itemid']), item)
    #        newentries += 1
    #        lastsync = item['eventtime']

    print "Fetching journal comments for: %s" % Journal

    try:
        f = open("%s/comment.meta" % Journal)
        metacache = pickle.load(f)
        f.close()
    except:
        metacache = {}

    try:
        f = open("%s/user.map" % Journal)
        usermap = pickle.load(f)
        f.close()
    except:
        usermap = {}

    maxid = lastmaxid
    while True:
        try:
            try:
                r = urllib2.urlopen(urllib2.Request(Server+"/export_comments.bml?get=comment_meta&startid=%d%s" % (maxid+1, authas), headers = {'Cookie': "ljsession="+ljsession}))
                meta = xml.dom.minidom.parse(r)
            except Exception, x:
                print "*** Error fetching comment meta, possibly not community maintainer?"
                print "***", x
                break
        finally:
            try:
                r.close()
            except AttributeError: # r is sometimes a dict for unknown reasons
                pass
        for c in meta.getElementsByTagName("comment"):
            id = int(c.getAttribute("id"))
            metacache[id] = {
                'posterid': c.getAttribute("posterid"),
                'state': c.getAttribute("state"),
            }
            if id > maxid:
                maxid = id
        for u in meta.getElementsByTagName("usermap"):
            usermap[u.getAttribute("id")] = u.getAttribute("user")
        if maxid >= int(meta.getElementsByTagName("maxid")[0].firstChild.nodeValue):
            break

    f = open("%s/comment.meta" % Journal, "w")
    pickle.dump(metacache, f)
    f.close()

    f = open("%s/user.map" % Journal, "w")
    pickle.dump(usermap, f)
    f.close()

    newmaxid = maxid
    maxid = lastmaxid
    while True:
        try:
            try:
                r = urllib2.urlopen(urllib2.Request(Server+"/export_comments.bml?get=comment_body&startid=%d%s" % (maxid+1, authas), headers = {'Cookie': "ljsession="+ljsession}))
                meta = xml.dom.minidom.parse(r)
            except Exception, x:
                print "*** Error fetching comment body, possibly not community maintainer?"
                print "***", x
                break
        finally:
            r.close()
        for c in meta.getElementsByTagName("comment"):
            id = int(c.getAttribute("id"))
            jitemid = c.getAttribute("jitemid")
            comment = {
                'id': str(id),
                'parentid': c.getAttribute("parentid"),
                'subject': gettext(c.getElementsByTagName("subject")),
                'date': gettext(c.getElementsByTagName("date")),
                'body': gettext(c.getElementsByTagName("body")),
                'state': metacache[id]['state'],
            }
            if usermap.has_key(c.getAttribute("posterid")):
                comment["user"] = usermap[c.getAttribute("posterid")]
            try:
                entry = xml.dom.minidom.parse("%s/C-%s" % (Journal, jitemid))
            except:
                entry = xml.dom.minidom.getDOMImplementation().createDocument(None, "comments", None)
            found = False
            for d in entry.getElementsByTagName("comment"):
                if int(d.getElementsByTagName("id")[0].firstChild.nodeValue) == id:
                    found = True
                    break
            if found:
                print "Warning: downloaded duplicate comment id %d in jitemid %s" % (id, jitemid)
            else:
                entry.documentElement.appendChild(createxml(entry, "comment", comment))
                f = codecs.open("%s/C-%s" % (Journal, jitemid), "w", "UTF-8")
                entry.writexml(f)
                f.close()
                newcomments += 1
            if id > maxid:
                maxid = id
        if maxid >= newmaxid:
            break

    lastmaxid = maxid

    writelast(Journal, lastsync, lastmaxid)

    if Username == Journal:
        print "Fetching userpics for: %s" % Username
        f = open("%s/userpics.xml" % Username, "w")
        print >>f, """<?xml version="1.0"?>"""
        print >>f, "<userpics>"
        for p in userpics:
            print >>f, """<userpic keyword="%s" url="%s" />""" % (p, userpics[p])
            pic = urllib2.urlopen(userpics[p])
            ext = MimeExtensions.get(pic.info()["Content-Type"], "")
            picfn = re.sub(r'[*?\\/:<>"|]', "_", p)
            try:
                picfn = codecs.utf_8_decode(picfn)[0]
                picf = open("%s/%s%s" % (Username, picfn, ext), "wb")
            except:
                # for installations where the above utf_8_decode doesn't work
                picfn = "".join([ord(x) < 128 and x or "_" for x in picfn])
                picf = open("%s/%s%s" % (Username, picfn, ext), "wb")
            shutil.copyfileobj(pic, picf)
            pic.close()
            picf.close()
        print >>f, "</userpics>"
        f.close()

    if origlastsync:
        print "%d new entries, %d new comments (since %s)" % (newentries, newcomments, origlastsync)
    else:
        print "%d new entries, %d new comments" % (newentries, newcomments)
    if errors > 0:
        print "%d errors" % errors

if __name__ == "__main__":
    if os.access("ljdump.config", os.F_OK):
        config = xml.dom.minidom.parse("ljdump.config")
        server = config.documentElement.getElementsByTagName("server")[0].childNodes[0].data
        username = config.documentElement.getElementsByTagName("username")[0].childNodes[0].data
        password = config.documentElement.getElementsByTagName("password")[0].childNodes[0].data
        journals = config.documentElement.getElementsByTagName("journal")
        if journals:
            for e in journals:
                ljdump(server, username, password, e.childNodes[0].data)
        else:
            ljdump(server, username, password, username)
    else:
        from getpass import getpass
        print "ljdump - livejournal archiver"
        print
        print "Enter your Livejournal username and password."
        print
        server = "http://livejournal.com"
        username = raw_input("Username: ")
        password = getpass("Password: ")
        print
        print "You may back up either your own journal, or a community."
        print "If you are a community maintainer, you can back up both entries and comments."
        print "If you are not a maintainer, you can back up only entries."
        print
        journal = raw_input("Journal to back up (or hit return to back up '%s'): " % username)
        print
        if journal:
            ljdump(server, username, password, journal)
        else:
            ljdump(server, username, password, username)
# vim:ts=4 et:	


ljdump-gui.py
#!/usr/bin/python
#
# ljdump-gui.py - gui interface to ljdump
# Greg Hewgill <greg@hewgill.com> http://hewgill.com
#
# NOTE: This is a work in progress and is probably not suitable for
#       general release just yet.
#
# LICENSE
#
# This software is provided 'as-is', without any express or implied
# warranty.  In no event will the author be held liable for any damages
# arising from the use of this software.
#
# Permission is granted to anyone to use this software for any purpose,
# including commercial applications, and to alter it and redistribute it
# freely, subject to the following restrictions:
#
# 1. The origin of this software must not be misrepresented; you must not
#    claim that you wrote the original software. If you use this software
#    in a product, an acknowledgment in the product documentation would be
#    appreciated but is not required.
# 2. Altered source versions must be plainly marked as such, and must not be
#    misrepresented as being the original software.
# 3. This notice may not be removed or altered from any source distribution.
#
# Copyright (c) 2005-2009 Greg Hewgill

import sys
import threading

from Tkinter import *

import ljdump

gWorkerThread = None

def poll():
    global gWorkerThread
    if gWorkerThread.isAlive():
        root.after(100, poll)
    else:
        gWorkerThread = None
        status['text'] = "Completed."
        ok['state'] = NORMAL
        cancel['state'] = NORMAL

def do_ok(event = None):
    print "ok"
    #root.withdraw()
    status['text'] = "Running..."
    ok['state'] = DISABLED
    cancel['state'] = DISABLED
    global gWorkerThread
    gWorkerThread = threading.Thread(None, ljdump.ljdump, args=("http://livejournal.com", username.get(), password.get()))
    gWorkerThread.start()
    poll()

def do_cancel(event = None):
    print "cancel", event
    root.destroy()

root = Tk()
root.title("ljdump")

body = Frame(root)
Label(body, text="Username:").grid(row=0, sticky=W)
Label(body, text="Password:").grid(row=1, sticky=W)
Label(body, text="Status:").grid(row=2, sticky=W)
username = Entry(body)
password = Entry(body, show="*")
status = Label(body, text="Waiting")
username.grid(row=0, column=1)
password.grid(row=1, column=1)
status.grid(row=2, column=1, sticky=W)
body.pack(padx=5, pady=5)

box = Frame(root)

ok = Button(box, text="OK", width=10, command=do_ok, default=ACTIVE)
ok.pack(side=LEFT, padx=5, pady=5)
cancel = Button(box, text="Cancel", width=10, command=do_cancel)
cancel.pack(side=LEFT, padx=5, pady=5)

root.bind("<Return>", do_ok)
root.bind("<Escape>", do_cancel)

box.pack()

username.focus_set()
root.mainloop()


convertdump.py
#!/usr/bin/python

# Copyright 2009, Sean M. Graham (www.sean-graham.com)
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
# 
# - Redistributions of source code must retain the above copyright notice,
#   this list of conditions and the following disclaimer.
# 
# - Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 

import xml.dom.minidom 
import os
import codecs
import sys
import getopt
import re

from time import strptime, strftime

def getNodeText(doc, nodename):
    rc = ""

    try:
        nodelist = doc.getElementsByTagName(nodename)[0].childNodes
    except:
        return ""

    for node in nodelist:
        if node.nodeType == node.TEXT_NODE:
            rc = rc + node.data

    return rc

def appendTextNode(doc, parent, nodename, value):
    nodeValue = value

    # make sure value is properly encoded
    try:
        bytes = nodeValue.encode("UTF-8")
    except:
        bytes = nodeValue.encode("cp1252")
        nodeValue = unicode(bytes, "UTF-8")

    element = doc.createElement(nodename)

    if( nodeValue != "" ): 
        textNode = doc.createTextNode(nodeValue)
        element.appendChild(textNode)

    parent.appendChild(element)


def addEntryForId(outDoc, element, username, id, includeSecure):
    entryFile = open("%s/L-%s" % (username,id), "r")
    inDoc = xml.dom.minidom.parse(entryFile)

    # Create an entry element
    entry = outDoc.createElement("entry")

    # Create an itemid element
    appendTextNode(outDoc, entry, "itemid", getNodeText(inDoc,"itemid"))

    # Create an eventtime element
    appendTextNode(outDoc, entry, "eventtime", getNodeText(inDoc, "eventtime"))

    # Create an subject element
    appendTextNode(outDoc, entry, "subject", getNodeText(inDoc, "subject"))

    # Create an event node (special case because for some reason there are two
    # 'event' elements in the pydump output, which is probably LJ's fault)
    event = inDoc.getElementsByTagName("event")[0]
    eventText = getNodeText(event, "event")

    appendTextNode(outDoc, entry, "event", replaceLJTags(eventText))

    security = getNodeText(inDoc, "security")

    if(security != ""):
        # don't append this entry unless the user provided the argument
        if(includeSecure == False):
            print("omitting secure entry: L-%s" % id)
            return 
        else:
            if(security == "usemask"):
                print("including allowmask entry: L-%s" % id)

                # Create an allowmask element 
                maskText = getNodeText(inDoc, "allowmask")

                if(maskText != ""):
                    appendTextNode(outDoc, entry, "allowmask", maskText)
                else:
                    appendTextNode(outDoc, entry, "allowmask", "0")
            else:
                print("including private entry: L-%s" % id)

        appendTextNode(outDoc, entry, "security", security)

    # Create a taglist element
    appendTextNode(outDoc, entry, "taglist", getNodeText(inDoc, "taglist"))

    # XXXSMG: make sure there is a comment file before trying to do anything
    # with it
    addCommentsForId(outDoc, entry, username, id)

    element.appendChild(entry)

def addCommentsForId(outDoc, entry, username, id):
    try: 
        commentFile = open("%s/C-%s" % (username,id), "r")
    except IOError:  # there are no comments for this entry
        return

    inDoc = xml.dom.minidom.parse(commentFile)

    comments = inDoc.getElementsByTagName("comment")

    for comment in comments:
        outComment = outDoc.createElement("comment")
        entry.appendChild(outComment)

        # add the item id for the comment
        appendTextNode(outDoc, outComment, "itemid", 
            getNodeText(comment, "id"))

        # convert the time string
        timeString = getNodeText(comment, "date")
        if( timeString != "" ):
            inDate = strptime(timeString, "%Y-%m-%dT%H:%M:%SZ")
            outDate = strftime("%Y-%m-%d %H:%M:%S", inDate)
            appendTextNode(outDoc, outComment, "eventtime", outDate)
        else:
            emptyTime = outDoc.createElement("eventtime")
            outComment.appendChild(emptyTime)

        # Create an subject element
        appendTextNode(outDoc, outComment, "subject", 
            getNodeText(comment, "subject"))

        # Create an event element
        bodyText = getNodeText(comment, "body")
        appendTextNode(outDoc, outComment, "event", replaceLJTags(bodyText))

        # Create the author element
        author = outDoc.createElement("author")
        outComment.appendChild(author)

        try:
            cUser = getNodeText(comment, "user")
        except:
            cUser = "anonymous"

        appendTextNode(outDoc, author, "name", cUser)
        appendTextNode(outDoc, author, "email", cUser + "@livejournal.com")
        
        # Create the parent_itemid
        parentId = getNodeText(comment, "parentid")
        if(parentId != ""): 
            appendTextNode(outDoc, outComment, "parent_itemid", parentId)


# regular expressions used in replaceLJTags()
#   (global for later reuse - suggestion by jparise)

userRE = re.compile('<lj user="(.*?)" ?/?>', re.IGNORECASE)
commRE = re.compile('<lj comm="(.*?)" ?/?>', re.IGNORECASE)
namedCutRE = re.compile('<lj-cut +text="(.*?)" ?/?>', 
                        re.IGNORECASE|re.DOTALL)
cutRE = re.compile('<lj-cut>', re.IGNORECASE)
cutRE = re.compile('</lj-cut>', re.IGNORECASE)
embedRE = re.compile('<lj-embed id="[0-9]+">', re.IGNORECASE)

def replaceLJTags(entry):
    rv = entry

    # replace lj user tags
    rv = re.sub(userRE, '<a href="http://www.livejournal.com/users/\\1" class="lj-user">\\1</a>', rv) 

    # replace lj comm tags
    rv = re.sub(commRE, '<a href="http://community.livejournal.com/\\1/" class="lj-comm">\\1</a>', rv) 

    # replace lj-cut tags
    rv = re.sub(namedCutRE, '<!--more \\1-->', rv)
    rv = re.sub(cutRE, '<!--more-->', rv)
    rv = re.sub(cutRE, '', rv)

    # replace lj-embed tags
    # this doesn't actually work.  LJ doesn't include the embedded content
    # when ljdump calls 'getevents', but instead includes an lj-embed tag
    # with an id and nothing else.
    #rv = re.sub(embedRE, '', rv)

    return rv


def usage():
    print( "Usage: convertdump.py [arguments]" )
    print( """
This will convert a pydump archive into something compatible with the
WordPress LiveJournal importer.  This is the same format used by the Windows
ljArchive exporter.

Arguments:
    -u  --user      username of archive to process [required]
    -l  --limit     limit the number of entries in each xml file (default 250)
    -i  --insecure  include private and protected entries in the output
    -h  --help      show this help page

Example:
    ./convertdump.py --user stevemartin --limit 200 --insecure
""")


def main(argv): 
    username = ""
    entryLimit = 250
    includeSecure = False;

    if( len(argv) == 0 ):
        usage()
        sys.exit(2)

    try:
        opts, args = getopt.getopt(sys.argv[1:], "hu:l:i", ["help",
                                                            "user=",
                                                            "limit=",
                                                            "insecure"])
    except getopt.GetoptError, err:
        # print help information and exit:
        print str(err) # will print something like "option -a not recognized"
        usage()
        sys.exit(2)

    for o, a in opts:
        if o == "-v":
            verbose = True
        elif o in ("-u", "--user"):
            username = a
        elif o in ("-l", "--limit"):
            entryLimit = int(a)
        elif o in ("-i", "--insecure"):
            print( "Warning:  Including secure entries in XML output" )
            includeSecure = True
        elif o in ("-h", "--help"):
            usage()
            sys.exit()
        else:
            assert False, "unhandled option"

    userDir = os.listdir(username)

    highNum = -1
    entryArray = []

    # get the list of entries
    for file in userDir:
        if file.startswith("L-"):
            entryNum = int(file.replace("L-",""))

            entryArray.append(entryNum)

            if( highNum < entryNum ):
                highNum = entryNum

    entryArray.sort()

    # Create the minidom document
    outDoc = xml.dom.minidom.Document()

    # Create the <livejournal> base element
    ljElement = outDoc.createElement("livejournal")
    outDoc.appendChild(ljElement)

    currentFileEntry = 0

    # start processing entries
    for entry in entryArray:
        addEntryForId(outDoc, ljElement, username, entry, includeSecure)

        currentFileEntry += 1

        if( currentFileEntry == entryLimit or entry == entryArray[-1] ):

            f = open("%s - %s.xml" % (username, entry), "w")
            tempXML = outDoc.toxml("UTF-8")
            f.write(tempXML)
            
            currentFileEntry = 0

            # Create the minidom document
            outDoc = xml.dom.minidom.Document()

            # Create the <livejournal> base element
            ljElement = outDoc.createElement("livejournal")
            outDoc.appendChild(ljElement)

if __name__ == "__main__":
    main(sys.argv[1:])

